Feature cart

hyun132 opened the pull request:
장바구니 기능 구현
-  CartRepository구현
- 상품번호 목록에 포함된 상품정보 읽어오는 기능 추가
- 상품 목록과 장바구니에 담은 갯수를 매핑해서 장바구니 목록 가져오는 기능 추가
- DataError타입의 오류 처리하기 위한 DataErrorHandler생성

hyun132 committed:
feature[#37] : 상품 번호 목록에 해당하는 상품목록을 가져오는 로직 추가

장바구니에 포함된 상품번호 목록에 해당하는 상품목록을 가져오는 로직 추가

hyun132 committed:
feature[#37] : CartRepository생성 및 장바구니 기능 관련 함수 정의

- CartRepository에 필요한 함수 정의
- 특정 유저의 장바구니에 저장된 상품번호 목록을 가져오는 기능 구현
- 목록의 데이터는 {상품번호 : 개수} 형태로 가져오도록 함

hyun132 committed:
feature[#37] : CartItem생성 및 GetCartListUseCase구현

- 상품번호 리스트에 해당하는 상품정보를 불러오는 로직 추가
- 상품정보와 장바구니에 추가된 갯수를 매핑
- DataError타입의 오류를 처리하는 Handler객체 생성

hyun132 committed:
feature[#37] : 장바구니 관련 usecase의 errorHandler타입을 DataErrorHandler로 변경

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/CartRepositoryImpl.kt:
```diff
@@ -0,0 +1,34 @@
+
+class CartRepositoryImpl @Inject constructor(
+    private val firestore: FirebaseFirestore,
+) : CartRepository<GetCartListUseCase.RequestParam> {
```
Repository가 UseCase의 파라미터에 의존되는 것이 아니라, UseCase가 Repository의 파라미터에 의존되는 형태로 가는 것이 아키텍처 구조상 올바른 방향일 것 같습니다.

hyun132 replied:
아 그러네요 수정하겠습니다!

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/CartRepositoryImpl.kt:
```diff
@@ -0,0 +1,34 @@
+        return firestore.collection("cart").document(requestParam.userId).snapshots()
+            .map { origin ->
+                val hashMap = HashMap<String, Long>()
+                origin.data?.keys?.forEach { key -> hashMap[key] = origin[key] as Long }
```
여기서 origin[key] 가 Long 타입이 아니라면 에러가 발생하고, 나머지 데이터들의 처리도 모두 중단 될 것 같습니다.
예외 처리를 한번 고민해보면 좋을 것 같습니다.

또한 HashMap 보다 명시적인 객체가 반환되는 것이 좋아보입니다.
map을 반환하게 되면, Presentation Layer에서 UseCase, Repository의 구현체가 내부에서 어떤 키들을 사용하는지 모두 알아야 하는데, 이는 유연하지 못한 구조가 될 가능성이 높습니다.

f-lab-dean commented on app/src/main/java/com/example/commerceapp/domain/usecases/cart/AddToCartUseCase.kt:
```diff
@@ -1,19 +1,17 @@
-    private val repository: CartRepository,
-    private val loginExceptionHandler: ErrorHandler
+    private val repository: CartRepository<RequestParam>,
+    private val dataErrorHandler: DataErrorHandler
```
에러 핸들러를 생성자의 의존성에 주입한 이유가 무엇일까요?
Hilt를 통해 에러핸들러가 주입되나요?

hyun132 replied:
넵 hilt로 핸들러를 주입하고 있습니다.
현재는 DataError타입을 사용하는 곳에서 모두 동일한 처리를 하고 있기 때문에 공통으로 사용하고 관리하려고 주입했습니다

f-lab-dean commented on app/src/main/java/com/example/commerceapp/domain/usecases/cart/GetCartListUseCase.kt:
```diff
@@ -1,19 +1,52 @@
+
+    private fun previewToCartItem(
+        productPreviews: List<ProductPreview>,
+        hashmap: HashMap<String, Long>
```
파라미터로 해시맵을 받는 것 보다, 명시적인 객체들을 받는 것이 좋습니다.
외부(Presentation Layer 등)에서 사용할 때, UseCase 또는 Repository가 어떤 키들을 사용하는지 구현 상세를 알아야 한다는 전제가 생길 수 있습니다.

hyun132 committed:
feature[#37] : [피드백 반영] Repository가 Usecase의 파라미터에 의존하는 부분 수정

Repository가 Usecase의 파라미터에 의존하는 부분을 GetCartListUseCase에서 repositoryImpl의 파라미터에 의존하도록 수정

hyun132 committed:
feature[#37] : [피드백 반영] 장바구니 목록을 나타내는 Cart객체 생성 및 CartRepository리턴 타입 HashMap에서 List<Cart>로 변경

- firestore의 데이터 구조 변경
- cart, CartDto, mapper객체 생성
- 장바구니 상품 목록을 HashMap에서 Cart로 변경
- getCartList의 리턴타입 HashMap에서 List<Cart>로 변경

hyun132 committed:
feature[#37] : cartItem 파일 cart디렉토리로 이동

hyun132 committed:
feature[#37] : 장바구니에 상품 추가, 삭제 기능 구현

- AddToCartUseCase, RemoveFromCartUseCase의 requestParam타입 변경,
- CartRepositoryImpl의 removeFromCart,addToCart 구현

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/CartRepositoryImpl.kt:
```diff
@@ -19,18 +21,72 @@ class CartRepositoryImpl @Inject constructor(
+        return flow {
+            val param = (requestParam as CartUpdateParam)
+            val collectionRef =
+                FirebaseFirestore.getInstance().collection("cart").document("testId")
```
Collection Key와 같이 고정적으로, 여러곳에서 사용되는 것들은 상수 등으로 만들어서 함께 사용될 수 있으면 좋습니다.

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/CartRepositoryImpl.kt:
```diff
@@ -19,18 +21,72 @@ class CartRepositoryImpl @Inject constructor(
-        TODO("Not yet implemented")
+    override suspend fun <CartRequestParam> addToCart(
+        requestParam: CartRequestParam
+    ): Flow<String> {
```
Add, Remove는 단발성 호출일 듯 한데 Flow를 사용한 이유가 있을까요?

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/CartRepositoryImpl.kt:
```diff
@@ -19,18 +21,72 @@ class CartRepositoryImpl @Inject constructor(
         requestParam: CartRequestParam
     ): Flow<List<Cart>> {
-        val param = requestParam as CartParam
+        val param = (requestParam as CartParam)
```
안정적인 형변환을 위해서 as? 를 사용하거나, 파라미터 자체를 CartParam로 받을 수 있도록 설계하는 것을 권장합니다.

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/CartRepositoryImpl.kt:
```diff
@@ -19,18 +21,72 @@ class CartRepositoryImpl @Inject constructor(
+        requestParam: CartUpdateParam
+    ): Flow<String> {
+        return flow {
+            val param = (requestParam as CartRepositoryImpl.CartUpdateParam)
```
requestParam은 이미 CartUpdateParam인데 변환해주어야 하는 이유가 무엇인가요?

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/CartRepositoryImpl.kt:
```diff
@@ -19,18 +21,72 @@ class CartRepositoryImpl @Inject constructor(
+        }
     }
 
     data class CartParam(
```
Param가 하위에 여러개 있는 것이라면 sealed interface를 활용하여 하나로 묶는 것은 어떠세요?
```
sealed interface CartParam {
    data class Update
    data class Add
    data class Remove
}
```
